File: jmlparser-jml-tools/src/main/java/com/github/jmlparser/xpath/AttribModifier.java
package com.github.jmlparser.xpath;

import com.github.javaparser.ast.Modifier;
import com.github.javaparser.ast.Node;
import com.github.javaparser.ast.nodeTypes.NodeWithModifiers;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.w3c.dom.Element;

import java.util.Arrays;
import java.util.Collection;

/**
 * @author Alexander Weigl
 * @version 1 (11.02.23)
 */
public class AttribModifier implements PseudoAttributeProvider {
    @Override
    public @Nullable Collection<JPAttrPseudo> attributeForNode(@NotNull Node node, @NotNull Element owner) {
        if (node instanceof NodeWithModifiers<?> m) {
            return Arrays.stream(Modifier.DefaultKeyword.values()).map(
                    it -> new JPAttrPseudo(it.name(), () -> "" + m.hasModifier(it), owner)).toList();
        }
        return null;
    }
}

File: tools/xpath/src/main/kotlin/io/github/jmltoolkit/xpath/AttribModifier.kt
package io.github.jmltoolkit.xpath

import com.github.javaparser.ast.Modifier.DefaultKeyword
import com.github.javaparser.ast.Node
import com.github.javaparser.ast.nodeTypes.NodeWithModifiers
import org.w3c.dom.Element

/**
 * @author Alexander Weigl
 * @version 1 (11.02.23)
 */
class AttribModifier : PseudoAttributeProvider {
    override fun attributeForNode(node: Node, owner: Element): Collection<JPAttrPseudo> {
        if (node is NodeWithModifiers<*>) {
            return DefaultKeyword.entries
                .map { JPAttrPseudo(it.name, { "" + node.hasModifier(it) }, owner) }
        }
        return listOf()
    }
}
